src/interpreter/interpreter.rs

```diff
@@ -4,22 +4,438 @@ use crate::ir::ast::Expression;
+        Box<Expression>,
+    ),
+    None,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum EvalResult {
+    CInt(i32),
+    CReal(f32),
+    Bool(bool),
+    List(Vec<EvalResult>),
+    None,
+}
 
-pub fn eval(exp: &Expression, env: &Environment) -> Result<IntValue, ErrorMessage> {
+type Environment = HashMap<Name, EnvValue>;
+
+pub fn eval(exp: &Expression, env: &Environment) -> Result<EvalResult, ErrorMessage> {
     match exp {
-        Expression::CInt(v) => Ok(*v),
-        Expression::Add(lhs, rhs) => Ok(eval(lhs, env)? + eval(rhs, env)?),
-        Expression::Sub(lhs, rhs) => Ok(eval(lhs, env)? - eval(rhs, env)?),
-        Expression::Mul(lhs, rhs) => Ok(eval(lhs, env)? * eval(rhs, env)?),
-        Expression::Div(lhs, rhs) => Ok(eval(lhs, env)? / eval(rhs, env)?),
+        Expression::CInt(v) => Ok(EvalResult::CInt(*v)),
+        Expression::CReal(v) => Ok(EvalResult::CReal(*v)),
+        Expression::Bool(v) => Ok(EvalResult::Bool(*v)),
+        Expression::None => Ok(EvalResult::None),
+        Expression::List(items) => {
+            let mut list_vec: Vec<EvalResult> = Vec::new();
+            let list_env = env.clone();
+
+            if items.len() < 1 {
+                return Err(String::from(
+                    "List initialization must have at least one element",
+                ));
+            } else {
+                let first_item = eval(&items[0], &list_env)?;
+                for item in items {
+                    let value = eval(&item, &list_env)?;
+                    match (&first_item, &value) {
+                        (EvalResult::CInt(_), EvalResult::CInt(_)) => list_vec.push(value),
+                        (EvalResult::CReal(_), EvalResult::CReal(_)) => list_vec.push(value),
+                        (EvalResult::Bool(_), EvalResult::Bool(_)) => list_vec.push(value),
+                        (EvalResult::List(_), EvalResult::List(_)) => list_vec.push(value),
+                        _ => return Err(String::from("List must be homogeneous")),
+                    }
+                }
+            }
+            Ok(EvalResult::List(list_vec))
+        }
+        Expression::Add(lhs, rhs) => {
+            let lhs_value = eval(lhs, env)?;
+            let rhs_value = eval(rhs, env)?;
+            match (lhs_value, rhs_value) {
+                (EvalResult::CInt(lhs), EvalResult::CInt(rhs)) => Ok(EvalResult::CInt(lhs + rhs)),
+                (EvalResult::CReal(lhs), EvalResult::CReal(rhs)) => {
+                    Ok(EvalResult::CReal(lhs + rhs))
+                }
+                (EvalResult::CInt(lhs), EvalResult::CReal(rhs)) => {
+                    Ok(EvalResult::CReal(lhs as f32 + rhs))
+                }
+                (EvalResult::CReal(lhs), EvalResult::CInt(rhs)) => {
+                    Ok(EvalResult::CReal(lhs + rhs as f32))
+                }
+                (EvalResult::CInt(lhs), EvalResult::Bool(rhs)) => {
+                    Ok(EvalResult::CInt(lhs + rhs as i32))
```

Comment:
Was it an explicit agreement to coerce bool to i32 in our language? I am aware that JavaScript will coerce the boolean value to an integer, and that `bool` is a subclass of `int` in Python, but do we want that in our language? I'm just asking because this may have implications for the type checker down the line.